libraries/plugins/elasticsearch/elasticsearch_plugin.cpp

```diff
@@ -0,0 +1,366 @@
+                     h = *o_op;
+               }));
+      };
+
+      if( !o_op.valid() ) {
+         _oho_index->use_next_id();
+         continue;
+      }
+      oho = create_oho();
+
+      const operation_history_object& op = *o_op;
+
+      // get the set of accounts this operation applies to
+      flat_set<account_id_type> impacted;
+      vector<authority> other;
+      operation_get_required_authorities( op.op, impacted, impacted, other ); // fee_payer is added here
+
+      if( op.op.which() == operation::tag< account_create_operation >::value )
+         impacted.insert( op.result.get<object_id_type>() );
+      else
+         graphene::app::operation_get_impacted_accounts( op.op, impacted );
+
+      for( auto& a : other )
+         for( auto& item : a.account_auths )
+            impacted.insert( item.first );
+
+      for( auto& account_id : impacted )
+      {
+         add_elasticsearch( account_id, oho->id );
+      }
+   }
+}
+
+void elasticsearch_plugin_impl::add_elasticsearch( const account_id_type account_id, const operation_history_id_type op_id )
+{
+   graphene::chain::database& db = database();
+   const auto &stats_obj = account_id(db).statistics(db);
+
+   // add new entry
+   const auto &ath = db.create<account_transaction_history_object>([&](account_transaction_history_object &obj) {
+      obj.operation_id = op_id;
+      obj.account = account_id;
+      obj.sequence = stats_obj.total_ops + 1;
+      obj.next = stats_obj.most_recent_op;
+   });
+
+   // keep stats growing as no op will be removed
+   db.modify(stats_obj, [&](account_statistics_object &obj) {
+      obj.most_recent_op = ath.id;
+      obj.total_ops = ath.sequence;
+   });
+
+   // curl buffers to read
+   std::string readBuffer;
+   std::string readBuffer_logs;
+
+   // ath data
+   std::string account_transaction = fc::json::to_string(ath.to_variant());
+
+   // operation history data
+   std::string trx_in_block = fc::json::to_string(op_id(db).trx_in_block);
+   std::string op_in_trx = fc::json::to_string(op_id(db).op_in_trx);
+
+   std::string operation_result = fc::json::to_string(op_id(db).result);
+   boost::replace_all(operation_result, "\"", "'");
+
+   std::string virtual_op = fc::json::to_string(op_id(db).virtual_op);
+
+   std::string op = fc::json::to_string(op_id(db).op);
+   // escaping some special characters inside op
+   boost::replace_all(op, "'", ""); // '
+   boost::replace_all(op, "\\'", ""); // \'
+   boost::replace_all(op, "\\\"", ""); // \"
+   boost::replace_all(op, "\"", "'"); // replace all " for '
+
+
+
+   std::string op_type = "";
+   if (!op_id(db).id.is_null())
+      op_type = fc::json::to_string(op_id(db).op.which());
+
+   std::string operation_history = "{\"trx_in_block\":" + trx_in_block + ",\"op_in_trx\":" + op_in_trx +
+                                   ",\"operation_results\":\"" + operation_result + "\",\"virtual_op\":" + virtual_op +
+                                   ",\"op\":\"" + op + "\"}";
```

Comment:
Please don't create JSON strings like that. At some point you'll forget to escape something, and then you're in trouble.
Use standard data structures (or maybe variant objects) together with fc::json::to_string instead.

Reply:
Strongly agree with this. The whole data conversion code (from an object to a parameter string to be passed into `cURL`) in this PR maybe work, but very "ugly" and error-prone. Rather than calling `to_string` then processing the strings, it's better to create the final strings directly from the objects. It's best if cURL and/or ElasticSearch can handle JSON directly (I think they should do so we just need to find out how), so we don't need to format the objects to another format.

Reply:
agree @abitmore . lot of stuff is ugly and have performance issues everywhere just as @pmconrad and you are pointing out. i will go over each one of them, was not expecting this pull to be accepted as it is just needed something working to be able to compare and get your comments. version1 actually did that job so now we just have to make it better.

Reply:
Thanks for your efforts.

Reply:
Yes! Please take everything I said as constructive suggestions. You did great.

Reply:
created some structures and making string at the end, removed all the ugly string creation in commit: https://github.com/bitshares/bitshares-core/pull/444/commits/95020fd77d491e5caf5396b748ba4bac8aedbf7a